fw/http_parser.c

```diff
@@ -8587,59 +8792,133 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			__FSM_EXIT(TFW_BLOCK);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("public"), &parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PUBLIC;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PUBLIC;
 		}, Resp_I_CC_p, Resp_I_Flag);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("private"), &parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PRIVATE;
-		}, Resp_I_CC_p, Resp_I_Flag);
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PRIVATE;
+		}, Resp_I_CC_p, Resp_I_Flag_Maybe_Qal);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("proxy-revalidate"),
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PROXY_REVAL;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PROXY_REVAL;
 		}, Resp_I_CC_p, Resp_I_Flag);
 		TRY_STR_INIT();
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	__FSM_STATE(Resp_I_Flag_Maybe_Qal) {
+		/* no-cache or private */
+		parser->cache_control.filled = true;
+		if (c == '=') {
+			parser->cache_control.dir_allowed = false;
+			/* Qualified form cancels out the general flag */
+			if (parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE)
+				resp->cache_ctl.flags |= TFW_HTTP_CC_PRIVATE_QUAL;
+			else
+				resp->cache_ctl.flags |= TFW_HTTP_CC_NO_CACHE_QUAL;
+
+			/* Resp_I_CC_Opening_Quote => Resp_I_CC_Flag_Argument =>
+			 * => Resp_I_EoT */
+			__FSM_I_MOVE_fixup(Resp_I_CC_Opening_Quote, 1, 0);
+		}
+		/* not qualified, fall though */
+	}
 	__FSM_STATE(Resp_I_Flag) {
-		WARN_ON_ONCE(!parser->_acc);
+		/* A start of a common flag successfully detected */
+		WARN_ON_ONCE(!parser->cache_control.dir_flag);
+		parser->cache_control.filled = true;
+
 		if (IS_WS(c) || c == ',' || IS_CRLF(c)) {
-			resp->cache_ctl.flags |= parser->_acc;
-			parser->_acc = 0;
+			/* Req_I_EoT will update the field at comma */
+			parser->cache_control.dir_allowed = false;
+			resp->cache_ctl.flags |= parser->cache_control.dir_flag;
 			__FSM_I_JMP(Resp_I_EoT);
 		}
+		/* ...but the flag appears to have an unknown suffix */
+		parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	__FSM_STATE(Resp_I_CC_Opening_Quote) {
+		if (c != '"')
+			__FSM_EXIT(TFW_BLOCK);
+		else
+			__FSM_I_MOVE_fixup(Resp_I_CC_Flag_Argument_Token, 1, 0);
+	}
+
+	__FSM_STATE(Resp_I_CC_Flag_Argument_Token) {
+		/* comma separated field list (field-name/token) with optional
+		 * linear white space */
+		TfwStrArray *tokens;
+		__FSM_I_MATCH_MOVE_fixup_finish(token, Resp_I_Ext, 0, {
+			tokens = parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE ?
+				 &resp->private_tokens : &resp->no_cache_tokens;
+			if (__str_array_add_token(msg, tokens,
+						  p, __fsm_sz, false))
+			{
+				__FSM_EXIT(TFW_BLOCK);
+			}
+		});
+
+		tokens = parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE ?
+			 &resp->private_tokens : &resp->no_cache_tokens;
+		if (__str_array_add_token(msg, tokens, p, __fsm_sz, false))
+			__FSM_EXIT(TFW_BLOCK);
+		/* Header name comparision requires a trailing semicolumn */
+		if (__str_array_add_token(msg, tokens, ":", 1, true))
+			__FSM_EXIT(TFW_BLOCK);
```

Comment:
Do not make a separate call with the generic allocation and TfwStr logic. You know precisely on the `__str_array_add_token()` call above that you finished to parse a header name and you can call the function with a flag, which will add the required `:`